README
Db_proxy version 1.08
=====================

This module and the included cgi can be used to proxy access to a database
via cgi-bin. This allows (for example) to share a single SQLite database
with multiple hosts.

This archive includes a client libarary used to access the installed
"server". It also includes the server which provides access to the database
to the clients and a sample client.

INSTALLATION

To install this module type the following:

   perl Makefile.PL [PREFIX=~]
   make
   make test
   make install

You will have to set up the 'server'. The bundled server is simple cgi with a 
plaintext configuration file. It contains no access controls so you will have
to use .htaccess or similar mechanisms to protect your database!
The usual *_proxy environment variables will be honoured.
If you want to access an SSL protected server see Crypt::SSLeay(3pm)
for relevant environment variables to configure things like 
client auth, ca-bundle, etc..

DEPENDENCIES

This module requires these other modules and libraries:

  HTTP::Request, YAML

If you want access SSL protected content you'll also need Crypt::SSLeay .

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2010 by Alexander Kuehn

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Db_proxy.

1.08  Tue Mar 16 2010
	- added http basic auth support

1.07  Tue Jul 28 2009
	- fixed get/post in initialization
	- fixed version number :>

1.06  Mon Jun 25 2009
	- added persistent connection to speed up subsequest requests and workaround Apache bug https://issues.apache.org/bugzilla/show_bug.cgi?id=12355 

1.05  Sat Feb 21 2009
	- add new function NewFetchTab, this was needed to be able to retain the column order

1.05  Fri Dec 17 2008
	- small changes to provide some error message
	- minor changes to the client to allow ENV variables in conf

1.04  Fri Oct 17 2008
	- switched from get to post to allow large queries

1.03  Thu Dec 20 2007
	- switched from Storable to YAML, too many problems with Storable
	- improved README and the server (new config setting to improve
	  campatibility with mod_perl)

1.02  Thu Dec 13 2007
	- fix compile error

1.01  Wed Dec 12 2007
	- use proxy if set
	- $Storable::interwork_56_64bit = 1; for better interoperability

1.00  Mon Dec  10 2007
	- added server and client

0.01  Tue Dec  4 2007
	- original version; created by h2xs 1.23 with options
		-X db_proxy



client/do.pl
#!/usr/bin/perl
#
# example client for Db_proxy.pm
# Alexander Kuehn <alexander.kuehn@hp.com>
# Example:
#    do.pl [configfile] "create table myTab(colA integer, colB text);" 
#

use strict;
use English;
use Db_proxy;
sub trim;

my $SCRIPT_NAME = $0;
# change currentdir to our dir so we can open the config
if ($0=~ /.+\//) {
	$SCRIPT_NAME = $POSTMATCH;
	chdir $MATCH if defined $MATCH;
}

local $::cfgfile = "db_proxy.cfg";	# default config file 
if ( defined $ARGV[0] && -r $ARGV[0] ) {
	$::cfgfile=shift @ARGV;
}
local $::config;

open (CFGFILE, $::cfgfile) or cgiDie ("Could not open $::cfgfile $!\n");
while (<CFGFILE>) {
	if ($_ =~ /^\s*\w+\s*=.+$/) {
		my ($key, $val)="";
		($key, $val) = split (/=/, $_, 2);
		trim($key, $val);
		chop( $val= eval "<<__EOD__\n$val\n__EOD__");	#expand variables
		if($key =~ /^ENV_/){
			$ENV{$POSTMATCH} = $val;
		} else {
			$::config{$key} = $val;
		}
	}
}
close CFGFILE;
if(defined $::config{"url"}) {
	$::config{"compression"}=(defined $::config{"compression"}) ? $::config{"compression"} : 0 ;
	my $db;
	if(defined $::config{"auth_realm"}) {
		$db = Db_proxy->new({ "url" => $::config{"url"}, "compression" => $::config{"compression"},
		 "auth_realm" => $::config{"auth_realm"}, "auth_user" => $::config{"auth_user"}, 
		 "auth_pass" => $::config{"auth_pass"} });
	} else {
		$db = Db_proxy->new({ url => $::config{"url"}, compression => $::config{"compression"} });
	}
	if($db) {
		my @res;
		my $res = $db->do(join ("\0", @ARGV), \@res);
		if ($res){
			print join("\n", @res) . "\n";
	}
} else {
	print "No url defined in $::cfgfile. :(\n";
}

exit(0);

##############################################################################
# function trims all parameters (eliminates whitespaces at beginning/end of strings)
# parameters: strings to be trimmed
# return: nothing - changes are made directly
##############################################################################
sub trim {
	for (@_) {
		if (defined $_){
			s/^\s+//;
			s/\s+$//;
		}
	}
}


client/fetch.pl
#!/usr/bin/perl
#
# example client for Db_proxy.pm
# Alexander Kuehn <alexander.kuehn@hp.com>
# Example:
#    fetch.pl [configfile] "select colA, colB from table;" | sed -n '1!p'
# the first line will be the column names which can filtered
# using sed as shown above
# It's also possible to specify several queries at once.
# The output will be appended in that case.
#
use strict;
use English;
use Db_proxy;
sub trim;

my $SCRIPT_NAME = $0;
# change currentdir to our dir so we can open the config
if ($0=~ /.+\//) {
	$SCRIPT_NAME = $POSTMATCH;
	chdir $MATCH if defined $MATCH;
}

local $::cfgfile = "db_proxy.cfg";	# default config file 
if ( defined $ARGV[0] && -r $ARGV[0] ) {
	$::cfgfile=shift @ARGV;		# if first arg is readable file assume its a configfile
}
local $::config;

open (CFGFILE, $::cfgfile) or die ("Could not open $::cfgfile $!\n");
while (<CFGFILE>) {
	if ($_ =~ /^\s*\w+\s*=.+$/) {
		my ($key, $val)="";
		($key, $val) = split (/=/, $_, 2);
		trim($key, $val);
		chop( $val= eval "<<__EOD__\n$val\n__EOD__");	#expand variables
		if($key =~ /^ENV_/){
			$ENV{$POSTMATCH} = $val;
		} else {
			$::config{$key} = $val;
		}
	}
}
close CFGFILE;
if(defined $::config{"url"}) {
	$::config{"compression"}=(defined $::config{"compression"}) ? $::config{"compression"} : 0 ;
	my $db;
	if(defined $::config{"auth_realm"}) {
		$db = Db_proxy->new({ "url" => $::config{"url"}, "compression" => $::config{"compression"},
		 "auth_realm" => $::config{"auth_realm"}, "auth_user" => $::config{"auth_user"}, 
		 "auth_pass" => $::config{"auth_pass"} });
	} else {
		$db = Db_proxy->new({ url => $::config{"url"}, compression => $::config{"compression"} });
	}
	if($db) {
		for (@ARGV) {
			my @res;
			my $res = $db->NewFetchTab($_, \@res);
			if ($res){
				for (my $cnt=0;$cnt<=$#res;$cnt++) {
					print join("\t", map { defined $_ ? $_ : "NULL" } @{$res[$cnt]}) . "\n";
				}
			}
		}
	}
} else {
	print "No url defined in $::cfgfile. :(\n";
}

exit(0);

##############################################################################
# function trims all parameters (eliminates whitespaces at beginning/end of strings)
# parameters: strings to be trimmed
# return: nothing - changes are made directly
##############################################################################
sub trim {
	for (@_) {
		if (defined $_){
			s/^\s+//;
			s/\s+$//;
		}
	}
}


server/cgi_lib.pm
# Perl Routines to Manipulate CGI input
# cgi-lib@pobox.com
#
# Copyright (c) 1993-1999 Steven E. Brenner  
# Unpublished work.
# Permission granted to use and modify this library so long as the
# copyright above is maintained, modifications are documented, and
# credit is given for any use of the library.
#
# Thanks are due to many people for reporting bugs and suggestions

# For more information, see:
#     http://cgi-lib.stanford.edu/cgi-lib/
package cgi_lib;
$cgi_lib'version = sprintf("%d.%02d", q$Revision: 2.3 $ =~ /(\d+)\.(\d+)/);


# Parameters affecting cgi_lib behavior
# User-configurable parameters affecting file upload.
$cgi_lib'maxdata    = 16777216;    # maximum bytes to accept via POST - 2^24
$cgi_lib'writefiles =      0;    # directory to which to write files, or
                                 # 0 if files should not be written
$cgi_lib'filepre    = "cgi_lib"; # Prefix of file names, in directory above

# Do not change the following parameters unless you have special reasons
$cgi_lib'bufsize  =  8192;    # default buffer size when reading multipart
$cgi_lib'maxbound =   100;    # maximum boundary length to be encounterd
$cgi_lib'headerout =    0;    # indicates whether the header has been printed


# ReadParse
# Reads in GET or POST data, converts it to unescaped text, and puts
# key/value pairs in %in, using "\0" to separate multiple selections

# Returns >0 if there was input, 0 if there was no input 
# undef indicates some failure.

# Now that cgi scripts can be put in the normal file space, it is useful
# to combine both the form and the script in one place.  If no parameters
# are given (i.e., ReadParse returns FALSE), then a form could be output.

# If a reference to a hash is given, then the data will be stored in that
# hash, but the data from $in and @in will become inaccessable.
# If a variable-glob (e.g., *cgi_input) is the first parameter to ReadParse,
# information is stored there, rather than in $in, @in, and %in.
# Second, third, and fourth parameters fill associative arrays analagous to
# %in with data relevant to file uploads. 

# If no method is given, the script will process both command-line arguments
# of the form: name=value and any text that is in $ENV{'QUERY_STRING'}
# This is intended to aid debugging and may be changed in future releases

sub ReadParse {
  # Disable warnings as this code deliberately uses local and environment
  # variables which are preset to undef (i.e., not explicitly initialized)
  local ($perlwarn);
  $perlwarn = $^W;
  $^W = 0;

  local (*in) = shift if @_;    # CGI input
  local (*incfn,                # Client's filename (may not be provided)
	 *inct,                 # Client's content-type (may not be provided)
	 *insfn) = @_;          # Server's filename (for spooled files)
  local ($len, $type, $meth, $errflag, $cmdflag, $got, $name);
	
  binmode(STDIN);   # we need these for DOS-based systems
  binmode(STDOUT);  # and they shouldn't hurt anything else 
  binmode(STDERR);
	
  # Get several useful env variables
  $type = $ENV{'CONTENT_TYPE'};
  $len  = $ENV{'CONTENT_LENGTH'};
  $meth = $ENV{'REQUEST_METHOD'};
  
  if ($len > $cgi_lib'maxdata) { #'
      &CgiDie("cgi_lib.pm: Request to receive too much data: $len bytes\n");
  }
  
  if (!defined $meth || $meth eq '' || $meth eq 'GET' || 
      $meth eq 'HEAD' ||
      $type eq 'application/x-www-form-urlencoded') {
    local ($key, $val, $i);
	
    # Read in text
    if (!defined $meth || $meth eq '') {
      $in = $ENV{'QUERY_STRING'};
      $cmdflag = 1;  # also use command-line options
    } elsif($meth eq 'GET' || $meth eq 'HEAD') {
      $in = $ENV{'QUERY_STRING'};
    } elsif ($meth eq 'POST') {
        if (($got = read(STDIN, $in, $len) != $len))
	  {$errflag="Short Read: wanted $len, got $got\n";};
    } else {
      &CgiDie("cgi_lib.pm: Unknown request method: $meth\n");
    }

    @in = split(/[&;]/,$in); 
    push(@in, @ARGV) if $cmdflag; # add command-line parameters

    foreach $i (0 .. $#in) {
      # Convert plus to space
      $in[$i] =~ s/\+/ /g;

      # Split into key and value.  
      ($key, $val) = split(/=/,$in[$i],2); # splits on the first =.

      # Convert %XX from hex numbers to alphanumeric
      $key =~ s/%([A-Fa-f0-9]{2})/pack("c",hex($1))/ge;
      $val =~ s/%([A-Fa-f0-9]{2})/pack("c",hex($1))/ge;

      # Associate key and value
      $in{$key} .= "\0" if (defined($in{$key})); # \0 is the multiple separator
      $in{$key} .= $val;
    }

  } elsif ($ENV{'CONTENT_TYPE'} =~ m#^multipart/form-data#) {
    # for efficiency, compile multipart code only if needed
$errflag = !(eval <<'END_MULTIPART');

    local ($buf, $boundary, $head, @heads, $cd, $ct, $fname, $ctype, $blen);
    local ($bpos, $lpos, $left, $amt, $fn, $ser);
    local ($bufsize, $maxbound, $writefiles) = 
      ($cgi_lib'bufsize, $cgi_lib'maxbound, $cgi_lib'writefiles);


    # The following lines exist solely to eliminate spurious warning messages
    $buf = ''; 

    ($boundary) = $type =~ /boundary="([^"]+)"/; #";   # find boundary
    ($boundary) = $type =~ /boundary=(\S+)/ unless $boundary;
    &CgiDie ("Boundary not provided: probably a bug in your server") 
      unless $boundary;
    $boundary =  "--" . $boundary;
    $blen = length ($boundary);

    if ($ENV{'REQUEST_METHOD'} ne 'POST') {
      &CgiDie("Invalid request method for  multipart/form-data: $meth\n");
    }

    if ($writefiles) {
      local($me);
      stat ($writefiles);
      $writefiles = "/tmp" unless  -d _ && -w _;
      # ($me) = $0 =~ m#([^/]*)$#;
      $writefiles .= "/$cgi_lib'filepre"; 
    }

    # read in the data and split into parts:
    # put headers in @in and data in %in
    # General algorithm:
    #   There are two dividers: the border and the '\r\n\r\n' between
    # header and body.  Iterate between searching for these
    #   Retain a buffer of size(bufsize+maxbound); the latter part is
    # to ensure that dividers don't get lost by wrapping between two bufs
    #   Look for a divider in the current batch.  If not found, then
    # save all of bufsize, move the maxbound extra buffer to the front of
    # the buffer, and read in a new bufsize bytes.  If a divider is found,
    # save everything up to the divider.  Then empty the buffer of everything
    # up to the end of the divider.  Refill buffer to bufsize+maxbound
    #   Note slightly odd organization.  Code before BODY: really goes with
    # code following HEAD:, but is put first to 'pre-fill' buffers.  BODY:
    # is placed before HEAD: because we first need to discard any 'preface,'
    # which would be analagous to a body without a preceeding head.

    $left = $len;
   PART: # find each part of the multi-part while reading data
    while (1) {
      die $@ if $errflag;

      $amt = ($left > $bufsize+$maxbound-length($buf) 
	      ?  $bufsize+$maxbound-length($buf): $left);
      $errflag = (($got = read(STDIN, $buf, $amt, length($buf))) != $amt);
      die "Short Read: wanted $amt, got $got\n" if $errflag;
      $left -= $amt;

      $in{$name} .= "\0" if defined $in{$name}; 
      $in{$name} .= $fn if $fn;

      $name=~/([-\w]+)/;  # This allows $insfn{$name} to be untainted
      if (defined $1) {
        $insfn{$1} .= "\0" if defined $insfn{$1}; 
        $insfn{$1} .= $fn if $fn;
      }
 
     BODY: 
      while (($bpos = index($buf, $boundary)) == -1) {
        if ($left == 0 && $buf eq '') {
	  foreach $value (values %insfn) {
            unlink(split("\0",$value));
	  }
	  &CgiDie("cgi_lib.pm: reached end of input while seeking boundary " .
		  "of multipart. Format of CGI input is wrong.\n");
        }
        die $@ if $errflag;
        if ($name) {  # if no $name, then it's the prologue -- discard
          if ($fn) { print FILE substr($buf, 0, $bufsize); }
          else     { $in{$name} .= substr($buf, 0, $bufsize); }
        }
        $buf = substr($buf, $bufsize);
        $amt = ($left > $bufsize ? $bufsize : $left); #$maxbound==length($buf);
        $errflag = (($got = read(STDIN, $buf, $amt, length($buf))) != $amt);
	die "Short Read: wanted $amt, got $got\n" if $errflag;
        $left -= $amt;
      }
      if (defined $name) {  # if no $name, then it's the prologue -- discard
        if ($fn) { print FILE substr($buf, 0, $bpos-2); }
        else     { $in {$name} .= substr($buf, 0, $bpos-2); } # kill last \r\n
      }
      close (FILE);
      last PART if substr($buf, $bpos + $blen, 2) eq "--";
      substr($buf, 0, $bpos+$blen+2) = '';
      $amt = ($left > $bufsize+$maxbound-length($buf) 
	      ? $bufsize+$maxbound-length($buf) : $left);
      $errflag = (($got = read(STDIN, $buf, $amt, length($buf))) != $amt);
      die "Short Read: wanted $amt, got $got\n" if $errflag;
      $left -= $amt;


      undef $head;  undef $fn;
     HEAD:
      while (($lpos = index($buf, "\r\n\r\n")) == -1) { 
        if ($left == 0  && $buf eq '') {
	  foreach $value (values %insfn) {
            unlink(split("\0",$value));
	  }
	  &CgiDie("cgi_lib: reached end of input while seeking end of " .
		  "headers. Format of CGI input is wrong.\n$buf");
        }
        die $@ if $errflag;
        $head .= substr($buf, 0, $bufsize);
        $buf = substr($buf, $bufsize);
        $amt = ($left > $bufsize ? $bufsize : $left); #$maxbound==length($buf);
        $errflag = (($got = read(STDIN, $buf, $amt, length($buf))) != $amt);
        die "Short Read: wanted $amt, got $got\n" if $errflag;
        $left -= $amt;
      }
      $head .= substr($buf, 0, $lpos+2);
      push (@in, $head);
      @heads = split("\r\n", $head);
      ($cd) = grep (/^\s*Content-Disposition:/i, @heads);
      ($ct) = grep (/^\s*Content-Type:/i, @heads);

      ($name) = $cd =~ /\bname="([^"]+)"/i; #"; 
      ($name) = $cd =~ /\bname=([^\s:;]+)/i unless defined $name;  

      ($fname) = $cd =~ /\bfilename="([^"]*)"/i; #"; # filename can be null-str
      ($fname) = $cd =~ /\bfilename=([^\s:;]+)/i unless defined $fname;
      $incfn{$name} .= (defined $in{$name} ? "\0" : "") . 
        (defined $fname ? $fname : "");

      ($ctype) = $ct =~ /^\s*Content-type:\s*"([^"]+)"/i;  #";
      ($ctype) = $ct =~ /^\s*Content-Type:\s*([^\s:;]+)/i unless defined $ctype;
      $inct{$name} .= (defined $in{$name} ? "\0" : "") . $ctype;

      if ($writefiles && defined $fname) {
        $ser++;
	$fn = $writefiles . ".$$.$ser";
	open (FILE, ">$fn") || &CgiDie("Couldn't open $fn\n");
        binmode (FILE);  # write files accurately
      }
      substr($buf, 0, $lpos+4) = '';
      undef $fname;
      undef $ctype;
    }

1;
END_MULTIPART
    if ($errflag) {
      local ($errmsg, $value);
      $errmsg = $@ || $errflag;
      foreach $value (values %insfn) {
        unlink(split("\0",$value));
      }
      &CgiDie($errmsg);
    } else {
      # everything's ok.
    }
  } else {
    &CgiDie("cgi_lib.pm: Unknown Content-type: $ENV{'CONTENT_TYPE'}\n");
  }

  # no-ops to avoid warnings
  $insfn = $insfn;
  $incfn = $incfn;
  $inct  = $inct;

  $^W = $perlwarn;

  return ($errflag ? undef :  scalar(@in)); 
}


# PrintHeader
# Returns the magic line which tells WWW that we're an HTML document

sub PrintHeader {
  return "Content-type: text/html\n\n";
}


# HtmlTop
# Returns the <head> of a document and the beginning of the body
# with the title and a body <h1> header as specified by the parameter

sub HtmlTop
{
  local ($title) = @_;

  return <<END_OF_TEXT;
<html>
<head>
<title>$title</title>
</head>
<body>
<h1>$title</h1>
END_OF_TEXT
}


# HtmlBot
# Returns the </body>, </html> codes for the bottom of every HTML page

sub HtmlBot
{
  return "</body>\n</html>\n";
}


# SplitParam
# Splits a multi-valued parameter into a list of the constituent parameters

sub SplitParam
{
  local ($param) = @_;
  local (@params) = split ("\0", $param);
  return (wantarray ? @params : $params[0]);
}


# MethGet
# Return true if this cgi call was using the GET request, false otherwise

sub MethGet {
  return (defined $ENV{'REQUEST_METHOD'} && $ENV{'REQUEST_METHOD'} eq "GET");
}


# MethPost
# Return true if this cgi call was using the POST request, false otherwise

sub MethPost {
  return (defined $ENV{'REQUEST_METHOD'} && $ENV{'REQUEST_METHOD'} eq "POST");
}


# MyBaseUrl
# Returns the base URL to the script (i.e., no extra path or query string)
sub MyBaseUrl {
  local ($ret, $perlwarn);
  $perlwarn = $^W; $^W = 0;
  $ret = 'http://' . $ENV{'SERVER_NAME'} .  
         ($ENV{'SERVER_PORT'} != 80 ? ":$ENV{'SERVER_PORT'}" : '') .
         $ENV{'SCRIPT_NAME'};
  $^W = $perlwarn;
  return $ret;
}


# MyFullUrl
# Returns the full URL to the script (i.e., with extra path or query string)
sub MyFullUrl {
  local ($ret, $perlwarn);
  $perlwarn = $^W; $^W = 0;
  $ret = 'http://' . $ENV{'SERVER_NAME'} .  
         ($ENV{'SERVER_PORT'} != 80 ? ":$ENV{'SERVER_PORT'}" : '') .
         $ENV{'SCRIPT_NAME'} . $ENV{'PATH_INFO'} .
         (length ($ENV{'QUERY_STRING'}) ? "?$ENV{'QUERY_STRING'}" : '');
  $^W = $perlwarn;
  return $ret;
}


# MyURL
# Returns the base URL to the script (i.e., no extra path or query string)
# This is obsolete and will be removed in later versions
sub MyURL  {
  return &MyBaseUrl;
}


# CgiError
# Prints out an error message which which containes appropriate headers,
# markup, etcetera.
# Parameters:
#  If no parameters, gives a generic error message
#  Otherwise, the first parameter will be the title and the rest will 
#  be given as different paragraphs of the body

sub CgiError {
  local (@msg) = @_;
  local ($i,$name);

  if (!@msg) {
    $name = &MyFullUrl;
    @msg = ("Error: script $name encountered fatal error\n");
  };

  if (!$cgi_lib'headerout) { #')
    print &PrintHeader;	
    print "<html>\n<head>\n<title>$msg[0]</title>\n</head>\n<body>\n";
  }
  print "<h1>$msg[0]</h1>\n";
  foreach $i (1 .. $#msg) {
    print "<p>$msg[$i]</p>\n";
  }

  $cgi_lib'headerout++;
}


# CgiDie
# Identical to CgiError, but also quits with the passed error message.

sub CgiDie {
  local (@msg) = @_;
  &CgiError (@msg);
  die @msg;
}


# PrintVariables
# Nicely formats variables.  Three calling options:
# A non-null associative array - prints the items in that array
# A type-glob - prints the items in the associated assoc array
# nothing - defaults to use %in
# Typical use: &PrintVariables()

sub PrintVariables {
  local (*in) = @_ if @_ == 1;
  local (%in) = @_ if @_ > 1;
  local ($out, $key, $output);

  $output =  "\n<dl compact>\n";
  foreach $key (sort keys(%in)) {
    foreach (split("\0", $in{$key})) {
      ($out = $_) =~ s/\n/<br>\n/g;
      $output .=  "<dt><b>$key</b>\n <dd>:<i>$out</i>:<br>\n";
    }
  }
  $output .=  "</dl>\n";

  return $output;
}

# PrintEnv
# Nicely formats all environment variables and returns HTML string
sub PrintEnv {
  &PrintVariables(*ENV);
}


# The following lines exist only to avoid warning messages
$cgi_lib'writefiles =  $cgi_lib'writefiles;
$cgi_lib'bufsize    =  $cgi_lib'bufsize ;
$cgi_lib'maxbound   =  $cgi_lib'maxbound;
$cgi_lib'version    =  $cgi_lib'version;
$cgi_lib'filepre    =  $cgi_lib'filepre;

1; #return true 



server/db_proxy.pl
#!/opt/perl_64/bin/bin/perl -w
#
# db proxy - the server portion to Db_proxy(3pm)
# provides access to a DBI datasource via plain HTTP(S)
# including gzip encoding support
# Author: Alexander Kuehn <alexander.kuehn@hp.com>
# 
#
# add include additional paths

unshift(@INC, '/home/user/public_html/cgi-bin/');

use strict;
use English;
use DBI;
use YAML qw(freeze);
require "cgi_lib.pm";
require "db_proxy_lib.pm";
local $::cfgfile = "/home/user/db/db_proxy.cfg";

#local $::True=1;
local $::False=0;

# hashs for config, input and array for owners
local %::config = ();

local (%::in,		# The form data, should be named %cgi_data
      %::cgi_cfn,   # The uploaded file(s) client-provided name(s)
      %::cgi_ct,    # The uploaded file(s) content-type(s).  These are
					#   set by the user's browser and may be unreliable
      %::cgi_sfn,	# The uploaded file(s) name(s) on the server (this machine)
      $::cgi_ret,	# Return value of the ReadParse call.
     );

local %::states = (
	"do"		=>	\&db_proxy_lib::st_do,
	"ft"		=>	\&db_proxy_lib::st_fetchtab,
	"fb"		=>	\&db_proxy_lib::st_newfetchtab
	);

&db_proxy_lib::readConfigFile;
# When writing files, several options can be set..
# Spool the files to the /tmp directory
$cgi_lib::writefiles = '/tmp';
# Limit upload size to avoid using too much memory
$cgi_lib::maxdata    = 16777216;
# The following lines are solely to suppress 'only used once' warnings
$cgi_lib::writefiles = $cgi_lib::writefiles;
$cgi_lib::maxdata    = $cgi_lib::maxdata;
$::cgi_ret = cgi_lib::ReadParse(\%::in,\%::cgi_cfn,\%::cgi_ct,\%::cgi_sfn);

local $::dbh = DBI->connect($::config{data_source}, $::config{dbuser}, $::config{dbpass});
local $::dberror;	#error string
unless ($::dbh) {
	print "$DBI::err ->\n$DBI::errstr";
	$::dberror=$DBI::errstr;
}
$::dbh->{RaiseError} = 1;
#$::dbh->commit;

#print "Content-Encoding: gzip\n" if ($ENV{"HTTP_ACCEPT_ENCODING"} =~ /gzip/);
print "Expires: -1\n"
	. "Pragma: no-cache\n"
	. "Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0\n";
print "Content-Type: application/x-yaml\n\n"; 

$|= 1; # - necessary for binary outputs
# gzip encoding
# set STDOUT to gzip pipe
#if ($ENV{"HTTP_ACCEPT_ENCODING"} =~ /gzip/) {
#	open (OUTPUT, "|$::config{gzip} -fc") || cgiDie ("Could not open gzip pipe!\n");
#	*STDOUT=*OUTPUT;
#}

if ($::states{$::in{"s"}}) {
	$::states{$::in{"s"}}->();
} else {
	cgiDie ("Invalid arguments!");
}
$::dbh->disconnect;
close OUTPUT;
exit;

 


lib/Db_proxy.pm
package Db_proxy;

use 5.006;
use strict;
use warnings;
use Carp;

use YAML qw(thaw);
require HTTP::Request;
require LWP::UserAgent;
require LWP::ConnCache;
require Exporter;
use AutoLoader qw(AUTOLOAD);

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Db_proxy ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '1.08';

sub new {
    my ( $class, $args ) = @_;
    $class = bless({}, $class);
	croak "Argument to new() must be a HASHREF" if ref $args ne 'HASH';
	for my $arg (qw( url )) {			# verify mandatory args
		if (exists $args->{$arg}) {
			$class->{$arg} = $args->{$arg};
        } else {
			croak "'$arg' argument required" if !exists $args->{$arg};
        }
	}
	my %defaults=( "compression"	=>	0,
					"auth_realm"	=>	"",
					"auth_user"		=>	"",
					"auth_pass"		=>	"");
	for my $arg (keys %defaults) {			# verify optional args
		if (exists $args->{$arg}) {
			$class->{$arg} = $args->{$arg};
        } else {
			$class->{$arg} = $defaults{$arg};
        }
	}
	$class->{'ua'} = LWP::UserAgent->new;
	$class->{'ua'}->env_proxy;
	$class->{'ua'}->timeout(15);
	$class->{'ua'}->conn_cache(LWP::ConnCache->new());
	if ($class->{'compression'}) {
		$class->{'ua'}->default_header("Accept-Encoding" => "gzip, deflate");
	}
	if ($class->{'auth_user'}) {
		my @t=split("/", $class->{'url'});
		unless($t[2] =~ /:\d+$/){		# unless a port is already given
			if($t[0] eq "https:"){
				$t[2].=":443";
			} else {
				$t[2].=":80";
			}
		}
		$class->{'ua'}->credentials($t[2], $class->{'auth_realm'}, 
			$class->{'auth_user'}, $class->{'auth_pass'});
	}
	# open & test the connection to workaround
	# https://issues.apache.org/bugzilla/show_bug.cgi?id=12355
	my $res = $class->{'ua'}->post($class->{'url'}, ['s' => 'ft', 'q' => 'select 1;']);
	if ($res->is_success) {
		return $class;
	} else {
		print STDERR $res->message;
		return undef;
	}
}

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Db_proxy - Perl extension for acessing a database via a HTTP(S) tunnel

=head1 SYNOPSIS

  use Db_proxy;
  my $db = Db_proxy->new( { url => 'https://foo.bar/cgi-bin/db_proxy.pl', compression => 1 } );
  my %res;
  my $res = $db->FetchTable("select * from table;", \%res);
  $db->do("delete from table where col like '';");


=head1 DESCRIPTION

This is a little wrapper library to access a centrally accessible database
through simple HTTP requests.
You need to install and configure the server before you can use this.
This has been tested with SQLite.

=head2 PARAMETERS

Mandatory:
 url	an URL as understood by HTTP::Request pointing to the db_proxy server

Optional:
 compression	gzip encoding, 1 = enable, 0 = disable, default = 0
 auth_realm		when using basic auth this is the real to authenticate against
 auth_user		the username to use for basic auth
 auth_pass		the password to use for basic auth

=head1 SEE ALSO

DBI(3pm), LWP(3pm), HTTP::Request(3pm), Crypt::SSLeay(3pm), Net::SSLeay(3pm)

=head1 AUTHOR

Alexander Kuehn, E<lt>nagilum@nagilum.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by Alexander Kuehn

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut

sub FetchTable {
	my $self = shift;
	my $q = shift;
	my $hash_ref = shift;
	my $res = $self->{'ua'}->post($self->{'url'}, ['s' => 'ft', 'q' => $q]);
	if ($res->is_success) {
		%{$hash_ref} = %{ thaw($res->decoded_content) };
	} else {
		$$hash_ref{"message"}=$res->message;
	}
	return $res->is_success;
}

sub do {
	my $self = shift;
	my $q = shift;
	my $ary_ref = shift;
	my $res = $self->{'ua'}->post($self->{'url'}, [ 's' => 'do', 'q' => $q]);
	if ($res->is_success) {
		@{$ary_ref} = @{ thaw($res->decoded_content) };
	} else {
		push @{$ary_ref}, $res->message;
	}
	return $res->is_success;
}

# fetch as a array of arrays
sub NewFetchTab {
	my $self = shift;
	my $q = shift;
	my $ary_ref = shift;
	my $res = $self->{'ua'}->post($self->{'url'}, ['s' => 'fb', 'q' => $q]);
	if ($res->is_success) {
		@{$ary_ref} = @{ thaw($res->decoded_content) };
	} else {
		$$hash_ref{"message"}=$res->message;
	}
	return $res->is_success;
}

